File: 2261-k-divisible-elements-subarrays/2261-k-divisible-elements-subarrays.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} p
 * @return {number}
 */
function TrieNode() {
    this.childs = new Map();
    this.isEnd = false;
}

var countDistinct = function (nums, k, p) {
    const root = new TrieNode();
    let cnt = 0, ans = 0, l = 0;
    const dfs = (curNode, i, end) => {
        if (i >= end) return;
        const val = nums[i];
        if (!curNode.childs.has(val)) {
            ++ans;
            const newNode = new TrieNode();
            newNode.isEnd = true;
            curNode.childs.set(val, newNode, end);
        }
        dfs(curNode.childs.get(val), i + 1, end);
    }
    for (let r = 0; r < nums.length; r++) {
        const val = nums[r];
        cnt += val % p ? 0 : 1;
        if (cnt <= k) continue;
        for (let j = l; j < r; ++j) dfs(root, j, r);
        while (cnt > k) {
            const num = nums[l++];
            cnt -= num % p ? 0 : 1;
        }
    }
    for (let j = l; j < nums.length; ++j) dfs(root, j, nums.length);
    return ans;
};

File: 2261-k-divisible-elements-subarrays/2261-k-divisible-elements-subarrays.cs
public class Solution
{
    List<int> a = new List<int>();
    int ans = 0;
    public class TrieNode
    {
        public Dictionary<int, TrieNode> childs = new Dictionary<int, TrieNode>();
        public bool isEnd;

        public TrieNode(bool isEnd = false)
        {
            this.isEnd = isEnd;
        }
    }
    public void dfs(TrieNode curNode, int i)
    {
        if (i >= a.Count) return;
        int val = a[i];
        if (!curNode.childs.ContainsKey(val))
        {
            ++ans;
            TrieNode newNode = new TrieNode(true);
            curNode.childs[val] = newNode;
        }
        dfs(curNode.childs[val], i + 1);
    }
    public int CountDistinct(int[] nums, int k, int p)
    {
        TrieNode root = new TrieNode();
        int cnt = 0;
        for (int i = 0; i < nums.Length; i++)
        {
            int val = nums[i];
            cnt += val % p != 0 ? 0 : 1;
            if (cnt > k)
            {
                for (int j = 0; j < a.Count; ++j) dfs(root, j);
                while (cnt > k)
                {
                    int num = a[0];
                    a.RemoveAt(0);
                    cnt -= num % p != 0 ? 0 : 1;
                }
            }
            a.Add(val);

        }
        for (int j = 0; j < a.Count; ++j) dfs(root, j);
        return ans;
    }
}